src/coom.jl

```diff
@@ -30,39 +30,56 @@ julia> using TextAnalysis, DataStructures
   [1, 2]  =  2.0
   [3, 2]  =  0.3999
   [2, 3]  =  0.3999
+
+julia> using TextAnalysis, DataStructures
+       doc = StringDocument("This is a text about an apple. There are many texts about apples.")
+       docv = TextAnalysis.tokenize(language(doc), text(doc))
+       vocab = OrderedDict("This"=>1, "is"=>2, "apple."=>3)
+       TextAnalysis.coo_matrix(Float16, docv, vocab, 5, true, :directional)
+
+3×3 SparseArrays.SparseMatrixCSC{Float16,Int64} with 4 stored entries:
+  [2, 1]  =  1.0
+  [1, 2]  =  1.0
+  [3, 2]  =  0.1999
+  [2, 3]  =  0.1999
 ```
 """
 function coo_matrix(::Type{T},
-                    doc::Vector{<:AbstractString},
-                    vocab::OrderedDict{<:AbstractString, Int},
-                    window::Int,
-                    normalize::Bool=true) where T<:AbstractFloat
+    doc::Vector{<:AbstractString},
+    vocab::OrderedDict{<:AbstractString,
+        Int},
+    window::Int,
+    normalize::Bool=true,
+    mode::Symbol=:default) where {T<:AbstractFloat}
+    # Initializations
     n = length(vocab)
     m = length(doc)
     coom = spzeros(T, n, n)
     # Count co-occurrences
     for (i, token) in enumerate(doc)
+        inner_range = if mode == :directional
+            i:min(m, i + window)
+        else
+            max(1, i - window):min(m, i + window)
+        end
         row = get(vocab, token, nothing)
-        isnothing(row) && continue
-
-        @inbounds for j in max(1, i - window):min(m, i + window)
-            i == j && continue
-
+        # looking forward
+        @inbounds for j in inner_range
             wtoken = doc[j]
-            col = get(vocab, wtoken, nothing)
-            isnothing(col) && continue
-
             nm = T(ifelse(normalize, abs(i - j), 1))
-            coom[row, col] += one(T) / nm
-            coom[col, row] = coom[row, col]
+            col = get(vocab, wtoken, nothing)
+            if i !== j && !isnothing(row) && !isnothing(col)
```

Comment:
It would be good to have `isnothing(row) && continue` outside the inner loop and avoid going into the loop.

And, actually, see the previous code.  `i == j && continue` previously allows to avoid doing something like `col = get(vocab, wtoken, nothing)`.

Reply:
Brilliant! I removed the if statement again. PS: I appeciate it that you took the time to explain it to an unexperienced user like me.

Reply:
Brilliant! I made the changes and I hope we are fine now (tests are passed). PS: Thanks for the taking the time to explain this to an unexperienced user like me.